### ISSUE TITLE
pre-start script is failed with FATAL:  database "registry" does not exist

### ISSUE THREAD

**Issue by tichimura-pivotal** [vmware/harbor-boshrelease#38]
Hi, we've been using harbor for PCF but upgrading Harbor from 1.6.0 to 1.6.3 has been failing for pre-start script.  You can see the logs in below.
As you can see, it is due to missing 'registry' database, and the script could not get the return of `get_version`[1], which should be returned from `get_version_pgsql` function [2] .
I wonder it should be recovered before, but does it mean failed to migration?

BTW, I'm wondering why this migration should not be needed for 1.6.0. This version is 1.6.0 but it still try to do the migration.  I wonder how the presence of vmware/harbor-log image[3] is the condition for this migration.
Let me know if you have any story behind for that.

Current Version: 1.6.0 (Harbor for PCF)
Target Version: 1.6.3 (Harbor for PCF)
Ops Manager: 2.3

Tomohiro

[1] https://github.com/goharbor/harbor/blob/v1.6.3/tools/migration/db/run.sh#L156
[2] https://github.com/goharbor/harbor/blob/v1.6.3/tools/migration/db/util/pgsql.sh#L133-L136
[3] https://github.com/vmware/harbor-boshrelease/blob/1.6/jobs/harbor/templates/bin/pre-start.erb#L185-L186

```
harbor-app/e951c164-18e3-445f-9fae-ec72d916f38f:/var/vcap/sys/log/harbor# cat pre-start.stdout.log
[Thu Jan 10 05:53:11 UTC 2019] Installing Harbor 1.6.3
loaded secret from file: /data/secretkey
Generated configuration file: /var/vcap/packages/harbor-app/common/config/nginx/nginx.conf
Generated configuration file: /var/vcap/packages/harbor-app/common/config/adminserver/env
Generated configuration file: /var/vcap/packages/harbor-app/common/config/ui/env
Generated configuration file: /var/vcap/packages/harbor-app/common/config/registry/config.yml
Generated configuration file: /var/vcap/packages/harbor-app/common/config/db/env
Generated configuration file: /var/vcap/packages/harbor-app/common/config/jobservice/env
Generated configuration file: /var/vcap/packages/harbor-app/common/config/jobservice/config.yml
Generated configuration file: /var/vcap/packages/harbor-app/common/config/log/logrotate.conf
Generated configuration file: /var/vcap/packages/harbor-app/common/config/registryctl/env
Generated configuration file: /var/vcap/packages/harbor-app/common/config/ui/app.conf
Copying UAA CA cert to /var/vcap/packages/harbor-app/common/config/ui/certificates/uaa_ca.pem
Generated certificate, key file: /var/vcap/packages/harbor-app/common/config/ui/private_key.pem, cert file: /var/vcap/packages/harbor-app/common/config/registry/root.crt
Copied custom ca bundle: /var/vcap/packages/harbor-app/common/config/custom-ca-bundle.crt
Copying sql file for notary DB
Generated certificate, key file: /var/vcap/packages/harbor-app/cert_tmp/notary-signer-ca.key, cert file: /var/vcap/packages/harbor-app/cert_tmp/notary-signer-ca.crt
Generated certificate, key file: /var/vcap/packages/harbor-app/cert_tmp/notary-signer.key, cert file: /var/vcap/packages/harbor-app/cert_tmp/notary-signer.crt
Copying certs for notary signer
Copying notary signer configuration file
Generated configuration file: /var/vcap/packages/harbor-app/common/config/notary/signer-config.postgres.json
Generated configuration file: /var/vcap/packages/harbor-app/common/config/notary/server-config.postgres.json
Copying nginx configuration file for notary
Generated configuration file: /var/vcap/packages/harbor-app/common/config/nginx/conf.d/notary.server.conf
loaded secret from file: /data/defaultalias
Generated configuration file: /var/vcap/packages/harbor-app/common/config/notary/signer_env
Generated configuration file: /var/vcap/packages/harbor-app/common/config/clair/postgres_env
Generated configuration file: /var/vcap/packages/harbor-app/common/config/clair/config.yaml
Generated configuration file: /var/vcap/packages/harbor-app/common/config/clair/clair_env
Create config folder: /var/vcap/packages/harbor-app/common/config/chartserver
Generated configuration file: /var/vcap/packages/harbor-app/common/config/chartserver/env
The configuration files are ready, please use docker-compose to start the service.
Client:
 Version:           18.06.0-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        0ffa825
 Built:             Wed Jul 18 19:04:39 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.0-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       0ffa825
  Built:            Wed Jul 18 19:13:39 2018
  OS/Arch:          linux/amd64
  Experimental:     false
[Thu Jan 10 05:54:01 UTC 2019] Docker daemon is running
[Thu Jan 10 05:54:01 UTC 2019] Loading docker images ...
Loaded image: goharbor/harbor-ui:v1.6.3
Loaded image: goharbor/harbor-jobservice:v1.6.3
Loaded image: goharbor/notary-signer-photon:v0.5.1-v1.6.3
Loaded image: goharbor/clair-photon:v2.0.6-v1.6.3
Loaded image: goharbor/nginx-photon:v1.6.3
Loaded image: goharbor/registry-photon:v2.6.2-v1.6.3
Loaded image: goharbor/notary-server-photon:v0.5.1-v1.6.3
Loaded image: goharbor/harbor-migrator:v1.6.3
Loaded image: goharbor/harbor-adminserver:v1.6.3
Loaded image: goharbor/harbor-log:v1.6.3
Loaded image: goharbor/harbor-db:v1.6.3
Loaded image: goharbor/redis-photon:v1.6.3
Loaded image: goharbor/chartmuseum-photon:v0.7.1-v1.6.3
[Thu Jan 10 05:54:14 UTC 2019] Upgrading Harbor 1.6.0 to 1.6.3 ...
[Thu Jan 10 05:54:14 UTC 2019] Use harbor-migrator:v1.6.3 for migration
[Thu Jan 10 05:54:14 UTC 2019] Backing up Harbor database
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locales
  COLLATE:  en_US.UTF-8
  CTYPE:    en_US.UTF-8
  MESSAGES: C
  MONETARY: C
  NUMERIC:  C
  TIME:     C
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory /var/lib/postgresql/data ... ok
creating subdirectories ... ok
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting dynamic shared memory implementation ... posix
creating configuration files ... ok
running bootstrap script ... ok
performing post-bootstrap initialization ... ok
syncing data to disk ... ok

Success. You can now start the database server using:

    pg_ctl -D /var/lib/postgresql/data -l logfile start

waiting for server to start....LOG:  database system was shut down at 2019-01-10 05:54:28 UTC
LOG:  MultiXact member wraparound protections are now enabled
LOG:  database system is ready to accept connections
LOG:  autovacuum launcher started
 done
server started
ALTER ROLE


./db/run.sh: ignoring /docker-entrypoint-initdb.d/*


PostgreSQL init process complete; ready for start up.

Performing backup...
TODO: needs to implement backup registry...
Backup performed.
harbor.cfg not found, WITH_ADMIRAL will not be set to true
[Thu Jan 10 05:54:30 UTC 2019] Migrating Harbor database
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locales
  COLLATE:  en_US.UTF-8
  CTYPE:    en_US.UTF-8
  MESSAGES: C
  MONETARY: C
  NUMERIC:  C
  TIME:     C
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory /var/lib/postgresql/data ... ok
creating subdirectories ... ok
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting dynamic shared memory implementation ... posix
creating configuration files ... ok
running bootstrap script ... ok
performing post-bootstrap initialization ... ok
syncing data to disk ... ok

Success. You can now start the database server using:

    pg_ctl -D /var/lib/postgresql/data -l logfile start

waiting for server to start....LOG:  database system was shut down at 2019-01-10 05:54:33 UTC
LOG:  MultiXact member wraparound protections are now enabled
LOG:  database system is ready to accept connections
LOG:  autovacuum launcher started
 done
server started
ALTER ROLE


./db/run.sh: ignoring /docker-entrypoint-initdb.d/*


PostgreSQL init process complete; ready for start up.

Version is not specified. Default version is head.
FATAL:  database "registry" does not exist
Please make sure to mount the correct the data volume.
harbor.cfg not found, WITH_ADMIRAL will not be set to true
```

**Comment by stonezdj** [comment]
Can you please check the content of /data/harbor_version in your harbor bosh vm before upgrade?
The file should content:
```1.6.0```

If this file is missed, then it will check the vmware/harbor-log to get the harbor version. and it might failed to upgrade.

**Comment by tichimura-pivotal** [comment]
Hi, yes, there is the content of "1.6.0", 
```
# cat /data/harbor_version
1.6.0
```
and I believe the "compareVersion" will return 0, but it seems to be checking vmware/harbor-log. 
https://github.com/vmware/harbor-boshrelease/blob/1.6/src/common/utils.sh

``` 
  searchResult=$( $DOCKER_CMD images | grep "vmware/harbor-log" || true )
  if [ -n "$searchResult" ] || [ $(compareVersion $INSTALLED_HARBOR_VERSION "1.6.0") = "-1" ]; then
...
```

**Comment by stonezdj** [comment]
You can login to the bosh vm and execute bash -x pre-start.sh to get detail trace of pre-start.sh execution

**Comment by stonezdj** [comment]
Fixed 1.7.2
